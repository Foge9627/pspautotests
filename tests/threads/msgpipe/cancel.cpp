#include "shared.h"

inline void testCancel(const char *title, SceUID msgpipe, bool doSend = true, bool doRecv = true) {
	int send = 1337, recv = 1337;
	int result = sceKernelCancelMsgPipe(msgpipe, doSend ? &send : NULL, doRecv ? &recv : NULL);
	if (result == 0) {
		checkpoint(NULL);
		schedf("%s: OK (sending=%d, receiving=%d) ", title, send, recv);
		schedfMsgPipe(msgpipe);
	} else {
		checkpoint("%s: Failed (%08x, sending=%d, receiving=%d)", title, result, send, recv);
	}
}

extern "C" int main(int argc, char *argv[]) {
	SceUID msgpipe = sceKernelCreateMsgPipe("cancel", PSP_MEMORY_PARTITION_USER, 0, 0x100, NULL);

	checkpointNext("Objects:");
	testCancel("  Normal", msgpipe);
	testCancel("  NULL", 0);
	testCancel("  Invalid", 0xDEADBEEF);
	testCancel("  Already canceled", msgpipe);
	sceKernelDeleteMsgPipe(msgpipe);
	testCancel("  Deleted", msgpipe);

	checkpointNext("Counts:");
	msgpipe = sceKernelCreateMsgPipe("cancel", PSP_MEMORY_PARTITION_USER, 0, 0x100, NULL);
	testCancel("  No send", msgpipe, false, true);
	testCancel("  No receive", msgpipe, true, false);
	testCancel("  Neither", msgpipe, false, false);
	sceKernelDeleteMsgPipe(msgpipe);

	checkpointNext("Waits:");
	{
		msgpipe = sceKernelCreateMsgPipe("Cancel", PSP_MEMORY_PARTITION_USER, 0, 0x100, NULL);
		MsgPipeReceiveWaitThread wait_r("receiving thread", msgpipe);
		testCancel("  With receiving thread", msgpipe);
		sceKernelDeleteMsgPipe(msgpipe);
	}

	{
		msgpipe = sceKernelCreateMsgPipe("delete", PSP_MEMORY_PARTITION_USER, 0, 0x100, NULL);
		// Send something to fill it up.
		char msg[256];
		sceKernelSendMsgPipe(msgpipe, msg, sizeof(msg), 0, NULL, NULL);
		MsgPipeSendWaitThread wait_s("sending thread", msgpipe);
		testCancel("  With sending thread", msgpipe);
		sceKernelDeleteMsgPipe(msgpipe);
	}

	return 0;
}