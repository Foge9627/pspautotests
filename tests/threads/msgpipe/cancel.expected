[x] Objects:
[x]   Normal: OK (sending=0, receiving=0) Msgpipe: OK (size=56, name=cancel, attr=00000000, buffer=100, free=100, sending=0, receiving=0)
[x]   NULL: Failed (8002019e, sending=1337, receiving=1337)
[x]   Invalid: Failed (8002019e, sending=1337, receiving=1337)
[x]   Already canceled: OK (sending=0, receiving=0) Msgpipe: OK (size=56, name=cancel, attr=00000000, buffer=100, free=100, sending=0, receiving=0)
[x]   Deleted: Failed (8002019e, sending=1337, receiving=1337)

[x] Counts:
[x]   No send: OK (sending=1337, receiving=0) Msgpipe: OK (size=56, name=cancel, attr=00000000, buffer=100, free=100, sending=0, receiving=0)
[x]   No receive: OK (sending=0, receiving=1337) Msgpipe: OK (size=56, name=cancel, attr=00000000, buffer=100, free=100, sending=0, receiving=0)
[x]   Neither: OK (sending=1337, receiving=1337) Msgpipe: OK (size=56, name=cancel, attr=00000000, buffer=100, free=100, sending=0, receiving=0)

[x] Waits:
[r]   ** started receiving thread
[x]   With receiving thread: OK (sending=0, receiving=1) Msgpipe: OK (size=56, name=Cancel, attr=00000000, buffer=100, free=100, sending=0, receiving=0)
[r]   ** receiving thread got result: 800201a9, received = 00000000
[r]   ** stopped receiving thread
[r]   ** started sending thread
[x]   With sending thread: OK (sending=1, receiving=0) Msgpipe: OK (size=56, name=delete, attr=00000000, buffer=100, free=100, sending=0, receiving=0)
[r]   ** sending thread got result: 800201a9, sent = 00000000
[r]   ** stopped sending thread
